### ISSUE TITLE
Separate ordered assertions from FakeScope

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#602]
Roughly arising from https://github.com/FakeItEasy/FakeItEasy/issues/484#issuecomment-184556236:

Fake Scopes, being implemented to rely on implicitly shared state, are problematic in a multi-threaded/async environment. We're examining improving or eliminating the Scopes.

One starting point would be to examine one of the biggest use of scopes, providing ordered assertions, and seeing if we can decouple that feature from the fake scopes.

Discuss!

**Comment by blairconrad** [comment]
First, let's see if we can come up with a good API. We'll need something that preserves the power of the existing ordered assertions, and that looks good. I'd recommend something that doesn't deviate too much from the syntax of our regular "unordered" assertions, if we can manage it, but let's see what happens.

**Comment by thomaslevesque** [comment]
Here's a recap of the proposals so far:

1.
```
Calls.To(
    () => x.Foo(42),
    () => y.Bar(),
    () => x.Baz())
    .MustHaveHappenedInOrder();
```
Pro: Simple and concise.
Con: No way to specify the repeat count. Can't be spread across xBehave steps.

2.
```
Calls.Ordered(a => 
{
    a.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once);
    a.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice);
    a.CallTo(() => x.Baz()).MustHaveHappened();
});
```
Pro: Similar to the familiar `A.CallTo(...)` syntax.
Con: The complete statement doesn't read like a sentence as other FIE APIs do. Can't be spread across xBehave steps.

3.
```
Ordered.Calls(to => 
{
    to(() => x.Foo(42)).Once();
    to(() => y.Bar()).Twice();
    to(() => x.Baz());
}).MustHaveHappened();
```
Pro: Reads like like a sentence
Con: Different from the usual assertion syntax. Can't be spread across xBehave steps.

4.
```
A
    .CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once)
    .AndThen
    .A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice)
    .AndThen
    .A.CallTo(() => y.Baz()).MustHaveHappened();
```
Pro: Reads like a sentence. Extends the current assertion syntax
Con: Can't be spread across xBehave steps. Probably more complex to implement.

5.
```
"Then x must have been foo'ed with 42 once"
    .x(() =>
    {
        a = new OrderedAssertions();
        a.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once);
    });

"And y must have been barred twice"
    .x(() => a.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice));

"Then y must have been bazzed"
    .x(() => a.CallTo(() => x.Baz()).MustHaveHappened());
```
Pro: Each assertion reads like a sentence. Can be spread across xBehave steps.
Con: Not sure it makes sense (more on that later)

**Comment by thomaslevesque** [comment]
About proposal 5:

> Con: Not sure it makes sense (more on that later)

I think we need to know all the assertions before we can check that they occurred in order. In this case, the first assertion checks that `x.Foo(42)` has been called once; but what if `x.SomethingElse()` was called before? Should it make the assertion fail? Because there might be another assertion later checking that `x.SomethingElse()` was called _after_ `x.Foo(42)`, but at this point we don't know it yet.

How is this currently handled with `scope.OrderedAssertions()`?

**Comment by blairconrad** [comment]
@thomaslevesque, we may be better off picking a favourite API first and then assessing for feasibility, but currently we walk through the recorded calls until we find a match for the call we're asserting. Then we _continue walking_ down the list or recorded calls to find the next thing to assert. And so on. So in that case we don't need to know the whole list of assertions up front. We just keep going until an assertion is not satisfied.

**Comment by thomaslevesque** [comment]
@blairconrad, ok, it makes sense. I withdraw my objection to proposal 5 then :wink:

**Comment by blairconrad** [comment]
So far I think I like **5** the best. But how do we feel about something even more verbose?

6. a.
```c#

new OrderedAssertions().Check(() => { // I'm not so sure about these names yet
    A.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once);
    A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice);
    A.CallTo(() => x.Baz()).MustHaveHappened();
});
```

Or, in a more xBehave-friendly format:


6. b.
```c#
"Then x must have been foo'ed with 42 once"
    .x(() =>
    {
        a = new OrderedAssertions();
        a.Check(() => A.CallTo(()=>x.Foo(42)).MustHaveHappened(Repeated.Once)));
    });

"And y must have been barred twice"
    .x(() => a.Check(() => A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice)));

"Then y must have been bazzed"
    .x(() => a.Check(()=> A.CallTo(() => x.Baz()).MustHaveHappened()));
```

In the first case, especially, I really enjoy how the `A.CallTo` syntax is not touched at all. I'm not wild about how 6.b. is a little more verbose, but could live with it, and as it's the less-likely use case, thought I'd put it out there.

**Comment by thomaslevesque** [comment]
I like 5 too, despite my initial objection. It has a nice syntax, and is flexible enough to spread the assertions across steps.

The 6a syntax is nice, but:
- The assertions can't be spread across steps
- The implementation would require ambient state, and #575 taught us to stay away from that. Of course, there would be no reason to make the lambda asynchronous, but there's nothing to prevent it either... Also, it would require the `A.CallTo(...).MustHaveHappened()` to be aware that it's being called in the context of an ordered assertion, which I don't like too much.

I don't really like 6b, too many nested lambdas make the code confusing and harder to read.

**Comment by blairconrad** [comment]
The 6.a. and 6.b. are actually the same syntax. It's just whether the included script block includes multiple statements and `Check` is called once, or whether it's called multiple times. So 6a spread across multiple steps is just 6b.

> The implementation would require ambient state
Nah. Check looks at the expression block.

> require A.CallTo(...).MustHaveHappened() to be aware that it's being called in the context of an ordered assertion

Unless we don't execute it. But even if it did, it knows _now_, so it's no worse.

**Comment by adamralph** [comment]
I like the fact that **6** leaves `A.CallTo()` unaltered. It's a shame that there's an extra nested lambda, but in the non-xbehave case that's really not much worse than the `using` block we have today.

In the xbehave case there's a nested lambda in each step, but that's the price for spreading the assertions across many steps.

I'm leaning towards 6 since it's the closest to what we have today, and is focused on only removing the ambient state and leaving everything else as-is. However,

> `Check` looks at the expression block.

Can all those statements really live in an expression block? And if it can, how easy is it to extract all of that out and analyse without executing it. If we executed it, surely we *would* need ambient state?

**Comment by blairconrad** [comment]
> > Check looks at the expression block.

> Can all those statements really live in an expression block?

Maybe!

> And if they can, how easy is it to extract all of that out and analyse without executing it?

Hard to say!

> If we executed it, surely we would need ambient state?

Perhaps. 

I'm sticking with "let's see what syntax we [really really want](http://www.azlyrics.com/lyrics/spicegirls/wannabe.html)" and then see if we can achieve it.

**Comment by thomaslevesque** [comment]
> Nah. Check looks at the expression block.

see next comment

> Can all those statements really live in an expression block?

No, they can't. The compiler can only generate expression trees for expressions, not for statements.

Just thought of another way to keep `A.CallTo()` unchanged (*), without introducing new lambdas:

```
using (var asserter = new OrderedAsserter()) // actual name to be defined
{
    A.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once, asserter);
    A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice, asserter);
    A.CallTo(() => x.Baz()).MustHaveHappened(asserter);
}
```
It keeps the familiar `A.CallTo()` syntax, and is xBehave-friendly (thanks to the `Using()` method, although I'm not sure it's really necessary to dispose the asserter)

Not sure the extra parameter in `MustHaveHappened` is the best approach. Perhaps something like this would be better:

    A.CallTo(() => x.Foo(42)).InOrder(asserter).MustHaveHappened(Repeated.Once);

_(*) personally  I don't find it very important to keep exactly `A.CallTo()`; I'd be fine with `a.CallTo()`, which is almost identical_

**Comment by blairconrad** [comment]
> The compiler can only generate expression trees for expressions, not for statements.

You're breaking my heart, @thomaslevesque.

> 
```c#
using (var asserter = new OrderedAsserter()) // actual name to be defined
{
    A.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once, asserter);
    A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice, asserter);
    A.CallTo(() => x.Baz()).MustHaveHappened(asserter);
}
```

I'd've called it inOrder:

```c#
using (var inorder = new OrderedAsserter()) // actual name to be defined
{
    A.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once, inOrder);
    A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice, inOrder);
    A.CallTo(() => x.Baz()).MustHaveHappened(inOrder);
}
```

**Comment by adamralph** [comment]
For 7a and 7b, there is no need for a `using` block:

```C#
var inOrder = new OrderedAsserter(); // actual name to be defined
A.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once, inOrder);
A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice, inOrder);
A.CallTo(() => x.Baz()).MustHaveHappened(inOrder);
```

7b is along the same lines as I suggested at https://github.com/FakeItEasy/FakeItEasy/issues/484#issuecomment-184556236

**Comment by thomaslevesque** [comment]
> I'd've called it inOrder:

Yes, that's probably better, but anyway the users will call it anyway they like ;)

What I like with this approach is that checking the order become just an extra option. The code is almost identical to normal assertions.

**Comment by thomaslevesque** [comment]
> For 7a and 7b, there is no need for a using block:

You're right, it's probably not necessary. On the other hand, if we realize later that we need to do some cleanup, suddenly making the asserter disposable would be a breaking change... Probably a case of YAGNI, though.

**Comment by adamralph** [comment]
Definitely YAGNI :smile:

**Comment by blairconrad** [comment]
> What I like with this approach is that checking the order become just an extra option. The code is almost identical to normal assertions.

Me too. I think I've new favourite. And it's somewhat discoverable. Once people are used to doing `MustHaveHappened`, they could stumble upon the "in order" extra parameter and know what to go looking for.

**Comment by adamralph** [comment]
Great, so it looks like we've killed off one reason to keep `FakeScope`.

**Comment by blairconrad** [comment]
Now that you point us back at https://github.com/FakeItEasy/FakeItEasy/issues/484#issuecomment-184556236, (e.g. 
`A.CallTo(() => fake.Bar(1)).In(scope).MustHaveHappened())`
)  I'm wondering a little more about using a new fluent term to point us at the "in orderness" instead of an extra param.

I don't think it reads well coming before `MustHaveHappened()`, though. Would we want to make the latter non-`void`?

```c#
var callOrder = new CallOrderAsserter(); // actual name to be defined
A.CallTo(() => fake.Bar(1)).MustHaveHappened().In(callOrder);
```

**Comment by thomaslevesque** [comment]
> I don't think it reads well coming before MustHaveHappened(), though.

I agree, it would be better to have it after

> Would we want to make the latter non-void?

Yes, I think so. `MustHaveHappened` would still check (immediately) that the method was called the specified number of times, and return "something" that would let the asserter check that the call happened in order.

**Comment by blairconrad** [comment]
Sounds good to me. The ordered asserter would have to redo the work of `MustHaveHappened`, but that shouldn't be a hardship.

**Comment by adamralph** [comment]
I swapped the `removal` label for the `breaking` label since we're not removing a feature (ordered assertions) but changing the way it is provided in a breaking manner.

**Comment by blairconrad** [comment]
Before we make this as **ready**, are we content with the syntax? I'm having a few last-minute misgivings. In particular, I worry that `In` doesn't really convey the whole story to the reader, and it only works because of the cutesy `callOrder` variable name. Perhaps it should be upgraded to `InOrder`, and the name of the parameter should matter less.
Maybe something a little more like

```c#
var callOrderContext = new CallOrderContext(); 
A.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once).InOrder(callOrderContext);
A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice).InOrder(callOrderContext);
A.CallTo(() => x.Baz()).MustHaveHappened().InOrder(callOrderContext);
```

? In this case, the user can call `callOrderContext` anything they want, if they value brevity.

As a second refinement, we could consider hanging `CallOrderContext` off `A`, to aid discoverability:

```c#
var callOrderContext = new A.CallOrderContext();
A.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once).InOrder(callOrderContext);
A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice).InOrder(callOrderContext);
A.CallTo(() => x.Baz()).MustHaveHappened().InOrder(callOrderContext);
```

**Comment by adamralph** [comment]
I like `InOrder()` and I like the `A` suggestion. I guess the `new` is superfluous, it could just be `var callOrderContext = A.CallOrderContext();`?

**Comment by thomaslevesque** [comment]
I like `InOrder` too, and I agree with @adamralph, the `new` isn't necessary. I can't think of any other place in FakeItEasy where the user has to explicitly call a constructor, so a static factory method is probably more consistent with the existing API.

**Comment by blairconrad** [comment]
Whoop `new` was a typo! Good catch.

**Comment by blairconrad** [comment]
I think we're **ready**.

**Comment by thomaslevesque** [comment]
> I think we're ready.

I agree. Unless someone else wants to do it, I think I'll take a stab at it this weekend.

**Comment by blairconrad** [comment]
If you're keen, I'll not stand in your way!

Have a Spec I was converting, from `OrderedCallMatchingSpecs.cs`. It's a work in progress, so no charge! Also, no obligation! Throw it all away. I won't be hurt!

```c#
        [Scenario]
        public static void NonGenericCalls(IHaveNoGenericParameters fake, ICallOrderingContext callOrderingContext, Exception exception)
        {
            "Given a Fake".x(() => fake = A.Fake<IHaveNoGenericParameters>());

            "And a call on the Fake, passing parameter 1".x(() => fake.Bar(1));

            "And a call on the Fake, passing parameter 2".x(() => fake.Bar(2));

            "And a call-ordering context".x(() => callOrderingContext = A.CallOrderingContext());

            "When I expect the call with parameter 2 to have been made".x(
                () => A.CallTo(() => fake.Bar(2)).MustHaveHappened(callOrderingContext));

            "And I expect the call with parameter 1 to have been made next".x(
                () => A.CallTo(() => fake.Bar(1)).MustHaveHappened(callOrderingContext));

            "Then the error should tell us that the calls were not matched in order".x(() =>
                                                                                       exception.Message.Should().Be(@"

  Assertion failed for the following calls:
    'FakeItEasy.Specs.OrderedCallMatchingSpecs+IHaveNoGenericParameters.Bar(2)' repeated at least once
    'FakeItEasy.Specs.OrderedCallMatchingSpecs+IHaveNoGenericParameters.Bar(1)' repeated at least once
  The calls where found but not in the correct order among the calls:
    1: FakeItEasy.Specs.OrderedCallMatchingSpecs+IHaveNoGenericParameters.Bar(baz: 1)
    2: FakeItEasy.Specs.OrderedCallMatchingSpecs+IHaveNoGenericParameters.Bar(baz: 2)
"));
```

**Comment by adamralph** [comment]
:+1:

**Comment by thomaslevesque** [comment]
Thanks for the spec @blairconrad!

**Comment by blairconrad** [comment]
@IharBury, [your suggestion](https://github.com/FakeItEasy/FakeItEasy/issues/484#issuecomment-185558833) of putting a "timestamp" on the calls was a key component of the fix.
Look for your name in the release notes. :trophy:

**Comment by matkoch** [comment]
I find it quite verbose tbh. And importantly the key thing is right at the end of those statements.

**Comment by blairconrad** [comment]
@matkoch, always a pleasure.

Can you explain what you mean bye "fluentish"? How is the current syntax not fluent, and what would you propose to improve it? The one that was implemented was the best we could come up with after quite a bit of agonizing, and I think we'd all be quite interested if you've got something better.

**Comment by thomaslevesque** [comment]
btw, the API described in the issue and comments isn't the one we settled on eventually. The actual syntax is this:

            var context = A.SequentialCallContext();
            A.CallTo(() => foo.Bar()).MustHaveHappened().InOrder(context);
            A.CallTo(() => foo.Baz()).MustHaveHappened().InOrder(context);

**Comment by matkoch** [comment]
I'm not sure about *something better* to be honest. But a few of my thoughts:

1. Your preferred syntax currently is quite repetitive with `MustHaveHappened().InOrder(context)` written a lot of times.
2. I'm not a fan of the argument 'can be spread across steps', because you in fact spread the assertion across multiple logical assertions. I think it will lead to a lot of descriptions not describing, that the actual assertion takes ordering into account.
3. To be honest I don't understand why a simple locking object for ordered assertions was abandoned so quickly. Using the `using` block was quite and it didn't affect existing API. The only wrong thing was the possibility for multi-threaded issues.

**Comment by adamralph** [comment]
The multi-threaded context is quite an important one these days. Parallel testing is the default in at least xunit 2.0 and people are going to want to use it. Turning it off in order to use FIE isn't really good enough. Add to that async tests (again, something people are going to want to use) and that basically eliminates ambient state as an option. Ambient state is a nasty anti-pattern and I'm glad we've gotten rid of it. Yes, it means you have to pass in a context object (explicit, rather than ambient, state) but I do think this is less of an evil than ambient state.

The argument for spreading assertions across steps is wider than just that. With the old `using` pattern, the assertions were queued up and only executed during the disposal. A using block does not display these semantics. It is a pattern for resource disposal and this is (was) one of the many misuses of it we see in the .NET world. The semantics which are communicated by a using block are, create some resource hogging object, do some things, release the resources. Anyone reading these semantics will be tripped up by the misuse, an author of an xbehave scenario being just one case.

**Comment by matkoch** [comment]
How are turning off parallel test execution if the code in question only runs a few milliseconds? That's what I assume, since it is only call validation code which is executed.

**Comment by adamralph** [comment]
The speed at which the tests run is irrelevant. We've already had reports form users that FIE 1.x intermittently breaks during parallel test execution and fake scope is the culprit.

**Comment by matkoch** [comment]
What exactly breaks? Sorry, don't know about that.

**Comment by adamralph** [comment]
@matkoch if you have two tests, both of which use the 1.x fake scope in order to record fake calls and then make assertions on the order of those calls, and you run those two tests in parallel a few times, eventually the ambient (static) state contained in the fake scope of one test will interfere with the other, and the ordered assertions will fail, because the recorded calls from one test will be intertwined the recorded calls from the other.

**Comment by matkoch** [comment]
Well, that's clear. And that's why I suggested a lock object.

**Comment by adamralph** [comment]
@matkoch perhaps it's better if we move this to https://gitter.im/FakeItEasy/FakeItEasy ?

**Comment by adamralph** [comment]
I guess we need to remove this from the 2.0 milestone and add `wontfix` once #637 is closed, since that will revert this feature?

**Comment by blairconrad** [comment]
Sounds good.
